# Bitacora: calidad de la secuenciación masiva y ensamblaje 

En esta bitácora se utilizarán diferentes programas 

# FastQC
Se puede encontrar más información en el sitio de [fastqc](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) 
<img src="img/fastqc1.png">
y de  [Trim Galore](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/)
<img src="./img/trim_galore.png">

# Ensamblaje con Soapdenovo2
Luo, R. et al. 2012. SOAPdenovo2: an empirically improved memory-efficient short-read de novo assembler. Gigascience 1, 18. [doi: 10.1186/2047-217X-1-18](https://academic.oup.com/gigascience/article/1/1/2047-217X-1-18/2656146)
SOAPdenovo2 [manual](https://github.com/aquaskyline/SOAPdenovo2)

### Todos estos programas se pueden instalar en sistemas Linux o iOS desde [anaconda](https://anaconda.org/bioconda/soapdenovo2)

In [1]:
import os
from Bio import SeqIO, Entrez

In [2]:
cd ~/Desktop/bioinformatica_anotacion2019/data/sec_masiva/

/Users/migueldelrio/Desktop/bioinformatica_anotacion2019/data/sec_masiva


In [3]:
ls

8_S356_L001_R1_001.fastq        8_S356_L001_R2_001_fastqc.html
8_S356_L001_R1_001_fastqc.html  8_S356_L001_R2_001_fastqc.zip
8_S356_L001_R1_001_fastqc.zip   fastqc/
8_S356_L001_R2_001.fastq


In [7]:
%%bash
grep "^@" ../8_S356_L001_R1_001.fastq | wc -l
grep "^@" ../8_S356_L001_R2_001.fastq | wc -l


    6660
    6674


### Número de secuencias obtenidas:

|archivo|secuencias|
|-------|----------|
|8_S356_L001_R1_001.fastq|6660|
|8_S356_L001_R2_001.fastq|6674|

In [8]:
%%bash
time fastqc 8_S356_L001_R1_001.fastq 
time fastqc 8_S356_L001_R2_001.fastq

Analysis complete for 8_S356_L001_R1_001.fastq
Analysis complete for 8_S356_L001_R2_001.fastq


Started analysis of 8_S356_L001_R1_001.fastq
Approx 15% complete for 8_S356_L001_R1_001.fastq
Approx 30% complete for 8_S356_L001_R1_001.fastq
Approx 45% complete for 8_S356_L001_R1_001.fastq
Approx 60% complete for 8_S356_L001_R1_001.fastq
Approx 75% complete for 8_S356_L001_R1_001.fastq
Approx 90% complete for 8_S356_L001_R1_001.fastq

real	0m7.147s
user	0m6.473s
sys	0m0.723s
Started analysis of 8_S356_L001_R2_001.fastq
Approx 15% complete for 8_S356_L001_R2_001.fastq
Approx 30% complete for 8_S356_L001_R2_001.fastq
Approx 45% complete for 8_S356_L001_R2_001.fastq
Approx 60% complete for 8_S356_L001_R2_001.fastq
Approx 75% complete for 8_S356_L001_R2_001.fastq
Approx 90% complete for 8_S356_L001_R2_001.fastq

real	0m5.702s
user	0m5.058s
sys	0m0.475s


In [4]:
ls -lh

total 15128
-rw-r--r--@  1 migueldelrio  staff   3.2M Jun  5  2013 8_S356_L001_R1_001.fastq
-rw-r--r--@  1 migueldelrio  staff   232K Feb 26 09:33 8_S356_L001_R1_001_fastqc.html
-rw-r--r--   1 migueldelrio  staff   241K Feb 26 09:33 8_S356_L001_R1_001_fastqc.zip
-rw-r--r--@  1 migueldelrio  staff   3.2M Jun  5  2013 8_S356_L001_R2_001.fastq
-rw-r--r--@  1 migueldelrio  staff   243K Feb 26 09:33 8_S356_L001_R2_001_fastqc.html
-rw-r--r--   1 migueldelrio  staff   260K Feb 26 09:33 8_S356_L001_R2_001_fastqc.zip
drwxr-xr-x  10 migueldelrio  staff   320B Feb 26 09:48 fastqc/


In [5]:
os.makedirs('fastqc',exist_ok=True)

In [5]:
cd fastqc/

/Users/migueldelrio/Desktop/bioinformatica_anotacion2019/data/sec_masiva/fastqc


In [12]:
%%bash
time trim_galore --paired ../8_S356_L001_R1_001.fastq ../8_S356_L001_R2_001.fastq

1.18


No quality encoding type selected. Assuming that the data provided uses Sanger encoded Phred scores (default)

Path to Cutadapt set as: 'cutadapt' (default)
Cutadapt seems to be working fine (tested command 'cutadapt --version')


AUTO-DETECTING ADAPTER TYPE
Attempting to auto-detect adapter type from the first 1 million sequences of the first file (>> ../8_S356_L001_R1_001.fastq <<)

Found perfect matches for the following adapter sequences:
Adapter type	Count	Sequence	Sequences analysed	Percentage
Nextera	2	CTGTCTCTTATA	6652	0.03
Illumina	0	AGATCGGAAGAGC	6652	0.00
smallRNA	0	TGGAATTCTCGG	6652	0.00
Using Nextera adapter for trimming (count: 2). Second best hit was Illumina (count: 0)

Writing report to '8_S356_L001_R1_001.fastq_trimming_report.txt'

SUMMARISING RUN PARAMETERS
Input filename: ../8_S356_L001_R1_001.fastq
Trimming mode: paired-end
Trim Galore version: 0.5.0
Cutadapt version: 1.18
Quality Phred score cutoff: 20
Quality encoding type selected: ASCII+33
Adapter sequence: 'C

In [11]:
ls -lh 

total 12880
-rw-r--r--   1 migueldelrio  staff   1.9K Feb 26 09:36 8_S356_L001_R1_001.fastq_trimming_report.txt
-rw-r--r--   1 migueldelrio  staff   3.2M Feb 26 09:36 8_S356_L001_R1_001_val_1.fq
-rw-r--r--   1 migueldelrio  staff   2.4K Feb 26 09:36 8_S356_L001_R2_001.fastq_trimming_report.txt
-rw-r--r--   1 migueldelrio  staff   3.1M Feb 26 09:36 8_S356_L001_R2_001_val_2.fq
-rw-r--r--   1 migueldelrio  staff   1.1K Feb 26 09:46 config_soap_8_S356.txt
drwxr-xr-x  23 migueldelrio  staff   736B Feb 26 09:48 output_directory/
-rw-r--r--   1 migueldelrio  staff   344B Feb 26 09:48 soapdenovo2.submit


In [18]:
!head -80 8_S356_L001_R1_001.fastq_trimming_report.txt


SUMMARISING RUN PARAMETERS
Input filename: ../8_S356_L001_R1_001.fastq
Trimming mode: paired-end
Trim Galore version: 0.5.0
Cutadapt version: 1.18
Quality Phred score cutoff: 20
Quality encoding type selected: ASCII+33
Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)
Maximum trimming error rate: 0.1 (default)
Minimum required adapter overlap (stringency): 1 bp
Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp


This is cutadapt 1.18 with Python 3.7.1
Command line parameters: -f fastq -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA ../8_S356_L001_R1_001.fastq
Processing reads on 1 core in single-end mode ...
Finished in 0.13 s (20 us/read; 2.96 M reads/minute).

=== Summary ===

Total reads processed:                   6,652
Reads with adapters:                     1,627 (24.5%)
Reads written (passing filters):         6,652 (100.0%)

Total basepairs processed:     1,477,486 bp
Quality-trimmed:                   8,177 bp (0.6%)
T

In [19]:
!head -80 8_S356_L001_R2_001.fastq_trimming_report.txt


SUMMARISING RUN PARAMETERS
Input filename: ../8_S356_L001_R2_001.fastq
Trimming mode: paired-end
Trim Galore version: 0.5.0
Cutadapt version: 1.18
Quality Phred score cutoff: 20
Quality encoding type selected: ASCII+33
Adapter sequence: 'CTGTCTCTTATA' (Nextera Transposase sequence; auto-detected)
Maximum trimming error rate: 0.1 (default)
Minimum required adapter overlap (stringency): 1 bp
Minimum required sequence length for both reads before a sequence pair gets removed: 20 bp


This is cutadapt 1.18 with Python 3.7.1
Command line parameters: -f fastq -e 0.1 -q 20 -O 1 -a CTGTCTCTTATA ../8_S356_L001_R2_001.fastq
Processing reads on 1 core in single-end mode ...
Finished in 0.15 s (22 us/read; 2.75 M reads/minute).

=== Summary ===

Total reads processed:                   6,652
Reads with adapters:                     1,909 (28.7%)
Reads written (passing filters):         6,652 (100.0%)

Total basepairs processed:     1,484,091 bp
Quality-trimmed:                  51,380 bp (3.5%)
T

In [12]:
%%bash
grep "^@" 8_S356_L001_R1_001_val_1.fq | wc -l
grep "^@" 8_S356_L001_R2_001_val_2.fq | wc -l

    6650
    6664


### Número de secuencias obtenidas:

|archivo|secuencias|archivo-trim|recortadas|
|-------|----------|----------|----------|
|8_S356_L001_R1_001.fastq|6660|8_S356_L001_R1_001_val_1.fq|6650|
|8_S356_L001_R2_001.fastq|6674|8_S356_L001_R2_001_val_2.fq|6664|

Tomado del manual de [Biopython](http://biopython.org/DIST/docs/tutorial/Tutorial.html#htoc292)

In [3]:
def cuentasecuencias(arch1):
    count, total_len, maximo = 0, 0, 0,
    minimo = 100
    for rec in SeqIO.parse(arch1, "fastq"):
        count += 1
        total_len += len(rec.seq)
        if maximo< len(rec.seq):
            maximo = len(rec.seq)
        if minimo > len(rec.seq):
            minimo = len(rec.seq)
    return (count, total_len, maximo, minimo)

In [8]:
count1, count2, count3, count4 = 0, 0, 0, 0
total_len1, total_len2, total_len3, total_len4, = 0, 0, 0, 0
minimo1, minimo2, minimo3, minimo4 = 100, 100, 100, 100
maximo1, maximo2, maximo3, maximo4 = 0, 0, 0, 0
# cuenta secuencias en el primer archivo
archivo1 = "../8_S356_L001_R1_001.fastq"
count1, total_len1, maximo1, minimo1 = cuentasecuencias(archivo1)
archivo2 = "8_S356_L001_R1_001_val_1.fq"
count2, total_len2, maximo2, minimo2 = cuentasecuencias(archivo2)
archivo3 = "../8_S356_L001_R2_001.fastq"
count3, total_len3, maximo3, minimo3 = cuentasecuencias(archivo3)
archivo4 = "8_S356_L001_R2_001_val_2.fq"
count4, total_len4, maximo4, minimo4 = cuentasecuencias(archivo4)

print ("archivo              \t#sec \tmin max \t archivo-trim\t\t\t #recor\tmin max")
print(archivo1[:17], "\t", count1, "\t", minimo1, maximo1, "\t", archivo2, "\t", count2, "\t", minimo2, maximo2)
print(archivo3[:17], "\t", count3, "\t", minimo3, maximo3, "\t", archivo4, "\t", count4, "\t", minimo4, maximo4)


archivo              	#sec 	min max 	 archivo-trim			 #recor	min max
../8_S356_L001_R1 	 6652 	 35 251 	 8_S356_L001_R1_001_val_1.fq 	 6642 	 24 251
../8_S356_L001_R2 	 6652 	 35 251 	 8_S356_L001_R2_001_val_2.fq 	 6642 	 21 251


# Archivo de configuración para el ensamblador Soapdenovo
En caso de ser necesario, cambiar el valor de `max_rd_len=`, en este caso a **251**  
Revisar los demás parámetros, p. e. `rd_len_cutoff=`  
Cambie las líneas q1 y q2, en este caso a:  
`q1=8_S356_L001_R1_001_val_1.fq  
q2=8_S356_L001_R2_001_val_2.fq`

Leer el [manual](https://vcru.wisc.edu/simonlab/bioinformatics/programs/soap/SOAPdenovo2MANUAL.txt).  
Asimismo es necesario tomar en cuenta el nombre del archivo de la configuración. En este caso:  
`config_soap_8_S356.txt`

In [25]:
fout = open("config_soap_8_S356.txt", "w")
linea="""#maximal read length
max_rd_len=251
[LIB]
#average insert size of the library
avg_ins=300
#if sequences are forward-reverse of reverse-forward
reverse_seq=0
#in which part(s) the reads are used (only contigs, only scaffolds, both contigs and scaffolds, only gap closure)
asm_flags=3
#cut the reads to the given length
rd_len_cutoff=300
#in which order the reads are used while scaffolding
rank=1
# cutoff of pair number for a reliable connection (at least 3 for short insert size)
pair_num_cutoff=3
#minimum aligned length to contigs for a reliable read location (at least 32 for short insert size)
map_len=32
#paired-end fastq files, read 1 file should always be followed by read 2 file
q1=8_S356_L001_R1_001_val_1.fq
q2=8_S356_L001_R2_001_val_2.fq
#another pair of paired-end fastq files, read 1 file should always be followed by read 2 file
#q1=input_reads2_pair_1.fq
#q2=input_reads2_pair_2.fq
#paired-end fasta files, read 1 file should always be followed by read 2 file
#f1= SRX5014491_1_val_1.fq
#f2=SRX5014491_2_val_2.fq
#fastq file for single reads
#q=input_reads.fq
"""

fout.write(linea)

fout.close()

### Verificando contenido del archivo de configuración

In [27]:
!head -50 config_soap_8_S356.txt

#maximal read length
max_rd_len=300
[LIB]
#average insert size of the library
avg_ins=300
#if sequences are forward-reverse of reverse-forward
reverse_seq=0
#in which part(s) the reads are used (only contigs, only scaffolds, both contigs and scaffolds, only gap closure)
asm_flags=3
#cut the reads to the given length
rd_len_cutoff=300
#in which order the reads are used while scaffolding
rank=1
# cutoff of pair number for a reliable connection (at least 3 for short insert size)
pair_num_cutoff=3
#minimum aligned length to contigs for a reliable read location (at least 32 for short insert size)
map_len=32
#paired-end fastq files, read 1 file should always be followed by read 2 file
q1=8_S356_L001_R1_001_val_1.fq
q2=8_S356_L001_R2_001_val_2.fq
#another pair of paired-end fastq files, read 1 file should always be followed by read 2 file
#q1=input_reads2_pair_1.fq
#q2=input_reads2_pair_2.fq
#paired-end fasta files, read 1 file should always be followed by read 2 file
#f1= SRX5014491_1_val_1.

In [ ]:
ls

# Archivo para ejecutar sh.
# Se debe cambiar el nombre del archivo generado, en este caso a:
`open("soapdenovo2.submit", "w")`

In [28]:
fout = open("soapdenovo2.submit", "w")
linea="""#!/bin/sh
#SBATCH --job-name=SOAPdenovo2
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=8
#SBATCH --time=168:00:00
#SBATCH --mem=50gb
#SBATCH --output=SOAPdenovo2.%J.out
#SBATCH --error=SOAPdenovo2.%J.err

#module load soapdenovo2/r240

SOAPdenovo-63mer all -s config_soap_8_S356.txt -K 31 -o output_directory/output31 -p $SLURM_NTASKS_PER_NODE


"""

fout.write(linea)

fout.close()

### Se verifica el contenido del archivo

In [29]:
!head -20 soapdenovo2.submit

#!/bin/sh
#SBATCH --job-name=SOAPdenovo2
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=8
#SBATCH --time=168:00:00
#SBATCH --mem=50gb
#SBATCH --output=SOAPdenovo2.%J.out
#SBATCH --error=SOAPdenovo2.%J.err

#module load soapdenovo2/r240

SOAPdenovo-63mer all -s config_soap_8_S356.txt -K 31 -o output_directory/output31 -p $SLURM_NTASKS_PER_NODE




### Se crea el directorio de salida, en caso de no existir.

In [30]:
os.makedirs('output_directory',exist_ok=True)

# Se ejecuta el programa

In [32]:
!time sh soapdenovo2.submit 


Version 2.04: released on July 13th, 2012
Compile Dec 23 2013	10:45:55
all: option requires an argument -- p

********************
Pregraph
********************

Parameters: pregraph -s config_soap_8_S356.txt -K 31 -o output_directory/output31 

In config_soap_8_S356.txt, 1 lib(s), maximum read length 300, maximum name length 256.

8 thread(s) initialized.
Import reads from file:
 8_S356_L001_R1_001_val_1.fq
Import reads from file:
 8_S356_L001_R2_001_val_2.fq
Time spent on hashing reads: 1s, 13284 read(s) processed.
LIB(s) information:
 [LIB] 0, avg_ins 300, reverse 0.
1879837 node(s) allocated, 2492602 kmer(s) in reads, 2492602 kmer(s) processed.
done hashing nodes
1855394 linear node(s) marked.
Time spent on marking linear nodes: 0s.
Time spent on pre-graph construction: 1s.

Start to remove frequency-one-kmer tips shorter than 62.
Total 6133 tip(s) removed.
8 thread(s) initialized.
1736 linear node(s) marked.
Start to remove tips with minority links.
437 tip(s) removed in cycle 1.

# Se verifican los archivos de salida y su contenido

In [7]:
ls output_directory/

output31.Arc               output31.preArc
output31.ContigIndex       output31.preGraphBasic
output31.bubbleInScaff     output31.readInGap.gz
output31.contig            output31.readOnContig.gz
output31.contigPosInscaff  output31.scaf
output31.edge.gz           output31.scafSeq
output31.gapSeq            output31.scafStatistics
output31.kmerFreq          output31.scaf_gap
output31.links             output31.updated.edge
output31.newContigIndex    output31.vertex
output31.peGrads


### Escriba en la siguiente celda qué contiene cada archivo.

### Verificando el contenido de *.contig

In [36]:
!head output_directory/output31.contig

>1 length 32 cvg_0.0_tip_0
AAAAAAGTACAGGCTCTGAAATGTACTCAAAG
>3 length 32 cvg_0.0_tip_0
AAAAATTAGAAACATCCTTTCTCAGAGTGATG
>5 length 32 cvg_0.0_tip_0
AAAACACACACACACACACACACACACACACA
>7 length 32 cvg_0.0_tip_0
AAAAGCTAATTCATGCATTTATTACTTCTAGG
>9 length 32 cvg_0.0_tip_0
AAAAGGGAGTTTTTCCTTCCCACTGTCACCAA


# Contando el número de contigs

In [43]:
!grep "^>" output_directory/output31.contig |wc -l

    8973


# Contando el número de secuencias en montaje (scaffolding)

In [44]:
!grep "^>" output_directory/output31.scafSeq |wc -l

    7834


# Estadística

In [45]:
!head -100 output_directory/output31.scafStatistics

<-- Information for assembly Scaffold 'output_directory/output31.scafSeq'.(cut_off_length < 100bp) -->

Size_includeN	1879435
Size_withoutN	1879435
Scaffold_Num	7834
Mean_Size	239
Median_Size	247
Longest_Seq	908
Shortest_Seq	100
Singleton_Num	7834
Average_length_of_break(N)_in_scaffold	0

Known_genome_size	NaN
Total_scaffold_length_as_percentage_of_known_genome_size	NaN

scaffolds>100 	7797	99.53
scaffolds>500 	21	0.27
scaffolds>1K  	0	0.00
scaffolds>10K 	0	0.00
scaffolds>100K	0	0.00
scaffolds>1M  	0	0.00

Nucleotide_A	553081	29.43
Nucleotide_C	402847	21.43
Nucleotide_G	387864	20.64
Nucleotide_T	535643	28.50
GapContent_N	0	0.00
Non_ACGTN	0	0.00
GC_Content	42.07		(G+C)/(A+C+G+T)

N10	408	424
N20	358	916
N30	314	1477
N40	251	2208
N50	250	2957
N60	248	3711
N70	234	4483
N80	203	5344
N90	160	6366

NG50	NaN	NaN
N50_scaffold-NG50_scaffold_length_difference	NaN

<-- Information for assembly Contig 'output_directory/output31.contig'.(cut_off_length < 100bp) -->

Size_includeN	1879435
Size_witho

## Una vez concluido el ensamblaje se procede a la anotación de los contigs